src/scheduler/sched_ifl_wrappers.cpp

```diff
@@ -0,0 +1,156 @@
+ * @param[in]	execvnode	-	the execvnode to run the job on
+ * @param[in]	node_owner	-	node owning server of first node in execvnode
+ *
+ * @return	int
+ * @retval	return value of the runjob call
+ */
+int
+send_run_job(int virtual_sd, int has_runjob_hook, char *jobid, char *execvnode,
+ 	     char *svr_id_node, char *svr_id_job)
+{
+	char extend[PBS_MAXHOSTNAME + 6];
+ 	int job_owner_sd;
+
+  	job_owner_sd = get_svr_inst_fd(virtual_sd, svr_id_job);
+
+  	extend[0] = '\0';
+ 	if (svr_id_node && svr_id_job && strcmp(svr_id_node, svr_id_job) != 0)
+ 		snprintf(extend, sizeof(extend), "%s=%s", SERVER_IDENTIFIER, svr_id_node);
+
+	if (sc_attrs.runjob_mode == RJ_EXECJOB_HOOK)
+		return pbs_runjob(job_owner_sd, jobid, execvnode, extend);
+	else if (((sc_attrs.runjob_mode == RJ_RUNJOB_HOOK) && has_runjob_hook))
+		return pbs_asyrunjob_ack(job_owner_sd, jobid, execvnode, extend);
+	else
+		return pbs_asyrunjob(job_owner_sd, jobid, execvnode, extend);
+}
+
+/**
+ * @brief
+ * 		send delayed attributes to the server for a job
+ *
+ * @param[in]	job_owner_sd	-	server connection descriptor of the job owner
+ * @param[in]	job_name	-	name of job for pbs_asyalterjob()
+ * @param[in]	pattr	-	attrl list to update on the server
+ *
+ * @return	int
+ * @retval	1	success
+ * @retval	0	failure to update
+ */
+int
+send_attr_updates(int job_owner_sd, char *job_name, struct attrl *pattr)
+{
+	const char *errbuf;
+	int one_attr = 0;
+
+	if (job_name == NULL || pattr == NULL)
+		return 0;
+
+	if (job_owner_sd == SIMULATE_SD)
+		return 1; /* simulation always successful */
+
+	if (pattr->next == NULL)
+		one_attr = 1;
+
+	if (pbs_asyalterjob(job_owner_sd, job_name, pattr, NULL) == 0) {
+		last_attr_updates = time(NULL);
+		return 1;
+	}
+
+	if (is_finished_job(pbs_errno) == 1) {
+		if (one_attr)
+			log_eventf(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_INFO, job_name,
+				   "Failed to update attr \'%s\' = %s, Job already finished",
+				   pattr->name, pattr->value);
+		else
+			log_event(PBSEVENT_SCHED, PBS_EVENTCLASS_JOB, LOG_INFO, job_name,
+				"Failed to update job attributes, Job already finished");
+		return 0;
+	}
+
+	errbuf = pbs_geterrmsg(job_owner_sd);
+	if (errbuf == NULL)
+		errbuf = "";
+	if (one_attr)
+		log_eventf(PBSEVENT_SCHED, PBS_EVENTCLASS_SCHED, LOG_WARNING, job_name,
+			   "Failed to update attr \'%s\' = %s: %s (%d)",
+			   pattr->name, pattr->value, errbuf, pbs_errno);
+	else
+		log_eventf(PBSEVENT_SCHED, PBS_EVENTCLASS_SCHED, LOG_WARNING, job_name,
+			"Failed to update job attributes: %s (%d)",
+			errbuf, pbs_errno);
+
+	return 0;
+}
+
+/**
+ * @brief	Wrapper for pbs_preempt_jobs
+ *
+ * @param[in]	virtual_sd - virtual sd for the cluster
+ * @param[in]	preempt_jobs_list - list of jobs to preempt
+ *
+ * @return	preempt_job_info *
+ * @retval	return value of pbs_preempt_jobs
+ */
+preempt_job_info *
+send_preempt_jobs(int virtual_sd, char **preempt_jobs_list)
+{
+    return pbs_preempt_jobs(virtual_sd, preempt_jobs_list);
+}
```

Comment:
Why do we need this function?

Reply:
I wanted to create IFL wrappers for the scheduler so that all of them could be found in one place. I had a hard time ensuring that all IFL calls from scheduler were correct as they were scattered throughout the code, so I thought that this will be useful. Let me know what you think.